backend/molgenis-emx2-tasks/src/main/java/org/molgenis/emx2/tasks/Task.java

```diff
@@ -0,0 +1,234 @@
+package org.molgenis.emx2.tasks;
+
+import static org.molgenis.emx2.tasks.TaskStatus.*;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.util.*;
+import java.util.function.Consumer;
+import org.molgenis.emx2.MolgenisException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * For documenting processes that consist of multiple steps and elements. For example. batch
+ * insert/upload tasks.
+ */
+public class Task implements Runnable, Iterable<Task> {
+  // some unique id
+  private String id = UUID.randomUUID().toString();
+  // for the toString method
+  private static ObjectMapper mapper =
+      new ObjectMapper().setSerializationInclusion(JsonInclude.Include.NON_NULL);
+  private Logger logger = LoggerFactory.getLogger(getClass().getSimpleName());
+  // human readable description
+  private String description;
+  // status of the tas
+  private TaskStatus status = WAITING;
+  // total size of the task, used for progress monitoring
+  private Integer total;
+  // position in the total size, if available, used for progress monitoring
+  private Integer progress;
+  // start time to measure run time
+  private long startTimeMilliseconds = System.currentTimeMillis();
+  // end time to calculate run time
+  long endTimeMilliseconds;
+  // subtasks/steps in this task
+  private List<Task> subTasks = new ArrayList<>();
+  // this parameter is used to indicate if steps should fail on unexpected state or should simply
+  // try to complete
+  private boolean strict = false;
+
+  public Task(String description) {
+    Objects.requireNonNull(description, "description cannot be null");
+    this.description = description;
+  }
+
+  public Task(String description, TaskStatus status) {
+    this(description);
+    Objects.requireNonNull(status, "status cannot be null");
+    this.status = status;
+  }
+
+  public Task(String description, boolean strict) {
+    this(description);
+    this.strict = strict;
+  }
+
+  public Task addSubTask(String message) {
+    Task step = new Task(message);
+    this.subTasks.add(step);
+    return step;
+  }
+
+  public Task addSubTask(String message, TaskStatus status) {
+    Task step = new Task(message, status);
+    this.subTasks.add(step);
+    return step;
+  }
+
+  public void addSubTask(Task task) {
+    Objects.requireNonNull(task, "task cannot be null");
+    this.subTasks.add(task);
+  }
+
+  public List<Task> getSubTasks() {
+    return subTasks;
+  }
+
+  public Integer getProgress() {
+    if (RUNNING.equals(status)) return progress;
```

Comment:
this.status ? , suggest to alway return the return type ( instead of null ) , should add a note about the range of the returnd int ( between 0 and 100 ? )